plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/SpringChangedHub.java

```diff
@@ -0,0 +1,267 @@
+            waitTimes.incrementAndGet();
+            if (pause.get()) {
+                LOGGER.trace("spring reload pause: {}", ObjectUtils.identityToString(current.beanFactory));
+                return;
+            }
+            if (status.compareAndSet(CHANGING, RELOADED)) {
+                // relead
+                try {
+                    current.reload();
+                } catch (Exception e) {
+                    LOGGER.error("reload spring failed: {}", e, ObjectUtils.identityToString(current.beanFactory));
+                }
+                finishReload();
+                waitTimes.set(0);
+            }
+        } else if (status.get() == CHANGING) {
+            if (waitTimes.get() == 0) {
+                int restTime = maxWaitTimes * DEFAULT_DELAY_PERIOD / 1000;
+                LOGGER.info("waiting to start reload '{}', it will start after {}s", ObjectUtils.identityToString(beanFactory()), restTime);
+            }
+            waitTimes.incrementAndGet();
+        }
+    }
+
+    public static void collectPlaceholderProperties(ConfigurableListableBeanFactory configurableListableBeanFactory) {
+        if (!(configurableListableBeanFactory instanceof DefaultListableBeanFactory)) {
+            return;
+        }
+        getInstance((DefaultListableBeanFactory) configurableListableBeanFactory).doCollectPlaceHolderProperties();
+    }
+
+    private void doCollectPlaceHolderProperties() {
+        String[] beanNames = defaultListableBeanFactory.getBeanDefinitionNames();
+        for (String beanName : beanNames) {
+            BeanDefinition beanDefinition = defaultListableBeanFactory.getBeanDefinition(beanName);
+            doCollectPlaceHolderProperties0(beanName, beanDefinition);
+        }
+    }
+
+    private void doCollectPlaceHolderProperties0(String beanName, BeanDefinition beanDefinition) {
+
+        if (beanDefinition.getPropertyValues() != null) {
+            for (PropertyValue pv : beanDefinition.getPropertyValues().getPropertyValues()) {
+                if (containPlaceHolder(pv.getValue(), beanName, beanDefinition)) {
+                    return;
+                }
+            }
+        }
+        if (beanDefinition.getConstructorArgumentValues().isEmpty()) {
+            return;
+        }
+        for (ConstructorArgumentValues.ValueHolder valueHolder : beanDefinition.getConstructorArgumentValues().getIndexedArgumentValues().values()) {
+            if (containPlaceHolder(valueHolder.getValue(), beanName, beanDefinition)) {
+                return;
+            }
+        }
+        for (ConstructorArgumentValues.ValueHolder valueHolder : beanDefinition.getConstructorArgumentValues().getGenericArgumentValues()) {
+            if (containPlaceHolder(valueHolder.getValue(), beanName, beanDefinition)) {
+                return;
+            }
+        }
+    }
+
+    private boolean containPlaceHolder(Object object, String beanName, BeanDefinition beanDefinition) {
```

Comment:
this method has side effect. I suggest to change it to 

```java
    private String getPlaceHolderBeanSource(Object object, BeanDefinition beanDefinition) {
        if (!isPlaceHolderBean(object)) {
            return null;
        }
        
        if (beanDefinition instanceof AbstractBeanDefinition) {
            return ResourceUtils.getPath(((AbstractBeanDefinition) beanDefinition).getResource());
        } else {
            return null;
        }
    }
```

Reply:
done